# Bronze Layer ETL
This script brings the data from the original .csv file into the previously created (empty) schema.

In [0]:
from pyspark.sql.functions import col, sum, isnan
from pyspark.sql.types import IntegerType 

In [0]:
# Saving the .csv file as a delta table
file_path = "/Workspace/Users/bru401@gmail.com/Main/Data_Engineering/ml_project1_data.csv"
#file_path = "dbfs:/Workspace/Users/bru401@gmail.com/Main/Data_Engineering/ml_project1_data.csv"

df = spark.read.csv(file_path, header=True, inferSchema=True)
df.write.format("delta").mode("overwrite").saveAsTable("bronze.raw_data")

---------------------------------------------------------------------------
SparkException                            Traceback (most recent call last)
File <command-8378329047444149>, line 6
      3 #file_path = "dbfs:/Workspace/Users/bru401@gmail.com/Main/Data_Engineering/ml_project1_data.csv"
      5 df = spark.read.csv(file_path, header=True, inferSchema=True)
----> 6 df.write.format("delta").mode("overwrite").saveAsTable("bronze.raw_data")

File /databricks/python/lib/python3.12/site-packages/pyspark/sql/connect/readwriter.py:737, in DataFrameWriter.saveAsTable(self, name, format, mode, partitionBy, **options)
    735 self._write.table_name = name
    736 self._write.table_save_method = "save_as_table"
--> 737 _, _, ei = self._spark.client.execute_command(
    738     self._write.command(self._spark.client), self._write.observations
    739 )
    740 self._callback(ei)

File /databricks/python/lib/python3.12/site-packages/pyspark/sql/connect/client/core.py:1556, in SparkConnectCli

### Alternatively

In [0]:
# Checking the schema for conflicts
print(spark.table("workspace.bronze.ml_project_2_data").printSchema())
print(spark.table("bronze.raw_data").printSchema())

root
 |-- ID: long (nullable = true)
 |-- Year_Birth: long (nullable = true)
 |-- Education: string (nullable = true)
 |-- Marital_Status: string (nullable = true)
 |-- Income: long (nullable = true)
 |-- Kidhome: long (nullable = true)
 |-- Teenhome: long (nullable = true)
 |-- Dt_Customer: date (nullable = true)
 |-- Recency: long (nullable = true)
 |-- MntWines: long (nullable = true)
 |-- MntFruits: long (nullable = true)
 |-- MntMeatProducts: long (nullable = true)
 |-- MntFishProducts: long (nullable = true)
 |-- MntSweetProducts: long (nullable = true)
 |-- MntGoldProds: long (nullable = true)
 |-- NumDealsPurchases: long (nullable = true)
 |-- NumWebPurchases: long (nullable = true)
 |-- NumCatalogPurchases: long (nullable = true)
 |-- NumStorePurchases: long (nullable = true)
 |-- NumWebVisitsMonth: long (nullable = true)
 |-- AcceptedCmp3: long (nullable = true)
 |-- AcceptedCmp4: long (nullable = true)
 |-- AcceptedCmp5: long (nullable = true)
 |-- AcceptedCmp1: long (nullab

In [0]:
# This block of code is used if the .csv file has been manually ingested (using its UI) into Databricks as a table.
df = spark.read.table("workspace.bronze.ml_project_2_data")
df.write.format("delta").mode("overwrite").saveAsTable("bronze.raw_data")

# Testing

In [0]:
display(df.limit(10))
print(f"The total number of lines is {df.count()}!")

ID Year_Birth Education Marital_Status Income Kidhome Teenhome Dt_Customer Recency MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain Z_CostContact Z_Revenue Response 5524 1957 Graduation Single 58138 0 0 2012-09-04 58 635 88 546 172 88 88 3 8 10 4 7 0 0 0 0 0 0 3 11 1 2174 1954 Graduation Single 46344 1 1 2014-03-08 38 11 1 6 2 1 6 2 1 1 2 5 0 0 0 0 0 0 3 11 0 4141 1965 Graduation Together 71613 0 0 2013-08-21 26 426 49 127 111 21 42 1 8 2 10 4 0 0 0 0 0 0 3 11 0 6182 1984 Graduation Together 26646 1 0 2014-02-10 26 11 4 20 10 3 5 2 2 0 4 6 0 0 0 0 0 0 3 11 0 5324 1981 PhD Married 58293 1 0 2014-01-19 94 173 43 118 46 27 15 5 5 3 6 5 0 0 0 0 0 0 3 11 0 7446 1967 Master Together 62513 0 1 2013-09-09 16 520 42 98 0 42 14 2 6 4 10 6 0 0 0 0 0 0 3 11 0 965 1971 Graduation Divorced 55635 0 1 2012-11-13 34 235 65 164 50 49 27 4 7 3 7 6 0 0 0 0 0 0 3 11 0 6177 1985 PhD Married 33454 1 0 2013-05-08 32 76 10 56 3 1 23 2 4 0 4 8 0 0 0 0 0 0 3 11 0 4855 1974 PhD Together 30351 1 0 2013-06-06 19 14 0 24 3 3 2 1 3 0 2 9 0 0 0 0 0 0 3 11 1 5899 1950 PhD Together 5648 1 1 2014-03-13 68 28 0 6 1 1 13 1 1 0 0 20 1 0 0 0 0 0 3 11 0

The total number of lines is 2240!


The total number of lines is 2240!

In [0]:
df.describe().display()

summary ID Year_Birth Education Marital_Status Income Kidhome Teenhome Recency MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain Z_CostContact Z_Revenue Response count 2240 2240 2240 2240 2216 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 2240 mean 5592.159821428571 1968.8058035714287 null null 52247.25135379061 0.44419642857142855 0.50625 49.109375 303.9357142857143 26.302232142857143 166.95 37.52544642857143 27.06294642857143 44.021875 2.325 4.084821428571429 2.6620535714285714 5.790178571428571 5.316517857142857 0.07276785714285715 0.07455357142857143 0.07276785714285715 0.06428571428571428 0.013392857142857142 0.009375 3.0 11.0 0.14910714285714285 stddev 3246.662197564341 11.984069456885825 null null 25173.07666090141 0.5383980977345874 0.5445382307698755 28.962452808378206 336.5973926053715 39.77343376457866 225.71537251175434 54.62897940287769 41.28049848785481 52.1674389149972 1.9322375008559614 2.7787141473881087 2.9231006555397197 3.250958145674417 2.426645009547285 0.25981306992189157 0.2627282848535636 0.2598130699218915 0.24531597433401 0.11497560625548155 0.0963911679444982 0.0 0.0 0.35627358640848406 min 0 1893 2n Cycle Absurd 1730 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 11 0 max 11191 1996 PhD YOLO 666666 2 2 99 1493 199 1725 259 263 362 15 27 28 13 20 1 1 1 1 1 1 3 11 1

Given that we know the number of lines is 2240, 'Income' having a count of only 2216 indicates that there are missing values.
This is the only column with such problem

In [0]:
# Checking the non-integer columns
display(
    df.groupBy("Education").count()
)

display(
    df.groupBy("Marital_Status").count()
)

Education count Graduation 1127 PhD 486 Master 370 Basic 54 2n Cycle 203

Marital_Status count Single 480 Together 580 Married 864 Divorced 232 Widow 77 Alone 3 Absurd 2 YOLO 2

Some of these categories seem incorrect and will be addressed when we clean the data in the silver layer.

# (WIP) Descriptive Analysis

In [0]:
display(df.dtypes)

_1 _2 ID bigint Year_Birth bigint Education string Marital_Status string Income bigint Kidhome bigint Teenhome bigint Dt_Customer date Recency bigint MntWines bigint MntFruits bigint MntMeatProducts bigint MntFishProducts bigint MntSweetProducts bigint MntGoldProds bigint NumDealsPurchases bigint NumWebPurchases bigint NumCatalogPurchases bigint NumStorePurchases bigint NumWebVisitsMonth bigint AcceptedCmp3 bigint AcceptedCmp4 bigint AcceptedCmp5 bigint AcceptedCmp1 bigint AcceptedCmp2 bigint Complain bigint Z_CostContact bigint Z_Revenue bigint Response bigint